Add support to traverse all python collection objects

ejguan opened the pull request:
Fixes https://github.com/pytorch/data/issues/752


This PR makes `traverse` function supporting more collections data structures from Python. The `getstate_hook` will be invoked after custom `__getstate__` function. This would guarantee that `traverse` function will be working as long as the `DataPipe` is working properly with multiprocessing.

ejguan commented:
@pytorchbot rebase

NivekT reviewed (commented):
Have you considered this instead? Or is this too broad?
```python
from collections.abc import Collection

isinstance(obj, Collection)  # Returns True for list, tuple, dict, set
```
The practical difference is that the above will cover any `Sequence` and also `ValuesView`. 

https://docs.python.org/3/library/collections.abc.html

ejguan commented:
> Have you considered this instead? Or is this too broad?
> 
> ```python
> from collections.abc import Collection
> 
> isinstance(obj, Collection)  # Returns True for list, tuple, dict, set
> ```
> 
> The practical difference is that the above will cover any `Sequence` and also `ValuesView`.
> 
> https://docs.python.org/3/library/collections.abc.html

Yeah. I did think about using `Iterable` from the `collections.abc`, which also covers all `list`, `tuple`, ... objects. And, I was thinking it would be too broad.

`Collection` seems like a better idea than `Iterable`. Thanks for your suggestion and I will change it.

ejguan commented:
@pytorchbot rebase

NivekT reviewed (approved):
LGTM!

ejguan commented:
@pytorchbot merge

weiwangmeta commented:
@pytorchbot revert -m="Diff reverted internally" -c="ghfirst"

Will re-merge this after smoothing out sync issues.

weiwangmeta commented:
@pytorchbot merge -l

weiwangmeta commented:
@pytorchbot merge

NivekT commented:
I think this was merged into nightly and broke some TorchData's TorchVision related CI tests.

https://github.com/pytorch/data/runs/8074482191?check_suite_focus=true#step:9:2359

Let's not land this for now.

cc: @VitalyFedyunin @ejguan

ejguan commented:
It turns out that `list` is the data type that potentially violates `traverse` function. The failure we have encountered in TorchVision is the buffer from `Demux` contains a list of file streams that can not be `pickled`.

NivekT reviewed (commented):
If it can happen to `list`, then I believe it can happen to any `Collection` since the object might contain unpickable streams.

Is it possible to always call `traverse` before the buffers of DataPipes are populated?
- It will require users to always call `apply_sharding`, `apply_shuffle_seed`, `apply_shuffle_settings` and etc before fill the DataPipes' buffer (through iteration or restoration).
- One thought came to mind is that we can save a copy of the graph in the beginning (maybe after applying adapter functions) and re-use that copy of the graph, but that would require no further insertion/removal of DataPipes. Not sure how feasible that is.

ejguan commented:
> Is it possible to always call `traverse` before the buffers of DataPipes are populated?

It might not be possible for torchvision's use case as they are loading some resources during the construction of DataPipe. I do have a workaround and will update the PR to see your opinion.

ejguan commented:
> I do have a workaround

The followings are two options in my mind.

1. Make collections of `DataPipes` with special classes. TorchData provides special form of `list`, `dict`, `tuple` that hold `DataPipes`. Then, `traverse` function would only take `DataPipe` or those special collection objects. It's easier for us to extract graph even during the iteration but it adds extra burden for DataPipe developers.
2. Convert `getstate_hook` to `getstate_wrapper`, which will rely on the implementation of `__getstate__` function from each `DataPipe`. `traverse` will only go through `DataPipe` or common python collection objects among the returned state from the original `__getstate__` function. The difference becomes the original `__getstate__` function should be able to handle unserialized buffer. Like `demux`, it currently ignores `buffer` during serialization and always recreate `buffer` during deserialization. However, the same problem still exists that `traverse` function might not be able to be called after `DataPipe`'s iteration.

ejguan commented:
Added a PR to guarantee this change won't break TorchVision: https://github.com/pytorch/vision/pull/6569

But, IMO, this **CAN NOT** be the ultimate solution for TorchData since we would require users to always do lazy loading even with their custom data pipeline. A proper solution will be updated in this PR.

ejguan commented:
Summarize the offline discussion with @VitalyFedyunin 

We agree on a comment that if the graph of DataPipe is working with Multiprocessing, it should also works with `traverse` function. So, our decision is to still rely on `hook`. But instead of calling hook before the custom `__getstate__` function, we will invoke hook directly over the `state` returned by `__getstate__` function. (using `self.__dict__` by default)

And, for future proofing, we might want to rely on wrapper over `__iter__` of `DataPipe` to always use a copy of `self` when it's not within a context of `DataLoader` in a single-process `DataLoader2`.

NivekT commented on torch/utils/data/graph.py:
```diff
@@ -74,6 +79,18 @@ def reduce_hook(obj):
+    ``list``, ``tuple``, ``set`` and ``dict``.
+    Args:
+        datapipe: the end DataPipe of the graph
+        only_datapipe: If ``False`` (default), all attributes of each DataPipe are traversed
```
nit: it seems like the default usages always set `only_datapipe=True`? Can you provide an example when you want it to be `False`? And perhaps mention that example in here for future reference?

ejguan replied:
This is for BC. We previously don't have `only_datapipe` option for `traverse` function. So, previously, this function behaves like `only_datapipe=False`.
I can add a deprecation warning here to indicate users to remove `only_datapipe` argument.

NivekT replied:
Yes I think a deprecation warning will be good if we do not expect it to be used by users in the future.

NivekT commented on torch/utils/data/graph.py:
```diff
@@ -28,24 +28,29 @@ def _list_connected_datapipes(scan_obj: DataPipe, only_datapipe: bool, cache: Se
             raise NotImplementedError
         else:
             captured_connections.append(obj)
-            cache.add(id(obj))
```
Question: Why is this removed? To allow visiting the same obj multiple times?

NivekT reviewed (commented):
Question: Does this change allow buffers to be non-empty when traversing the graph? If so, how?

NivekT commented:
> And, for future proofing, we might want to rely on wrapper over `__iter__` of `DataPipe` to always use a copy of `self` when it's not within a context of `DataLoader` in a single-process `DataLoader2`.

Sorry I missed the meeting, but what does this mean? We should create an isolated copy of the DataPipe when `__iter__` is called? So that the state is independent from the DataPipe object itself?

ejguan commented:
> Does this change allow buffers to be non-empty when traversing the graph? If so, how?

No, it won't. Rather than let `traverse` function handle non-empty buffer itself, we decided to make it relying on the implementation of `IterDataPipe.__getstate__`. All non-empty buffer should be specially handled within `__getstate__` function, as the same problem would happen for multiprocessing.

> Sorry I missed the meeting, but what does this mean? We should create an isolated copy of the DataPipe when `__iter__` is called? So that the state is independent from the DataPipe object itself?

As we have discussed before, when `DataPipe` is iterated by `DataLoader2`, it should be a copy of the original `DataPipe` graph being iterated. Then, for the case that `DataPipe` is iterated without using `DataLoader2`, to guarantee that the original `DataPipe` graph won't generate any `buffer`, a new copy of `DataPipe` graph should be used for the iteration of `DataPipe`.

ejguan commented:
> The reasoning makes sense but it might get tricky for `ChildDataPipe` cases. I'd imagine you want to stay on the same graph while iterating through two `ChildDataPipes` with the same parent. There might be a trivial solution but I will need to think more.

We might need to do another follow-up with Vitaly on this topic.

ejguan committed:
Add support to traverse all python collection objects (#773)

Summary:
X-link: https://github.com/pytorch/data/pull/773

Fixes https://github.com/pytorch/data/issues/752

This PR makes `traverse` function supporting more collections data structures from Python. The `getstate_hook` will be invoked after custom `__getstate__` function

Pull Request resolved: https://github.com/pytorch/pytorch/pull/84079

Differential Revision: D39357886

Pulled By: ejguan

fbshipit-source-id: 3c0629f7a355d2f7a67c7224a49b2b2721592778

huydhn commented:
@pytorchbot merge -g
